tests/src/ut_test_kvp.c

```diff
@@ -130,6 +140,159 @@ void test_ut_kvp_open( void )
+        fclose(file);
+        pInstance->length = 0;
+        return -1;
+    }
+
+    // Null-terminate the buffer
+    pInstance->buffer[pInstance->length] = '\0';
+
+    fclose(file);
+    return 0;
+}
+
+void test_ut_kvp_open_memory( void )
+{
+    ut_kvp_instance_t *pInstance = NULL;
+    ut_kvp_status_t status;
+    test_ut_memory_t pMemoryInstance;
+
+    /* Negative Test */
+    UT_LOG_STEP("ut_kvp_openMemory( NULL, NULL, -1 )");
+    status = ut_kvp_openMemory( NULL, NULL, -1);
+    UT_ASSERT( status == UT_KVP_STATUS_INVALID_INSTANCE );
+
+    /* Positive Test */
+    UT_LOG_STEP("ut_kvp_createInstance");
+    pInstance = ut_kvp_createInstance();
+    UT_ASSERT( pInstance != NULL );
+
+    /* Negative Test*/
+    UT_LOG_STEP("ut_kvp_close( pInstance ) - Not been opened");
+    ut_kvp_close(pInstance);
+
+    /* Negative Read Test, NULL PARAM */
+    UT_LOG_STEP("ut_kvp_openMemory( pInstance, NULL ) - Negative");
+    status = ut_kvp_openMemory( pInstance, NULL, -1);
+    UT_ASSERT( status == UT_KVP_STATUS_INVALID_PARAM );
+
+    /* data doesn't exist */
+    UT_LOG_STEP("ut_kvp_openMemory( pInstance, %s - filename doesn't exist ) - Negative", KVP_VALID_TEST_NO_FILE);
+    const char *kvp_str = "  ";
+    char* kvp_buffer = strdup(kvp_str);
+    if (kvp_buffer)
+    {
+        status = ut_kvp_openMemory(pInstance, kvp_buffer, -1);
+        UT_ASSERT(status == UT_KVP_STATUS_PARSING_ERROR);
+    }
+
+    /* Zero length file, that the library should reject because it can't parse it at all */
+    UT_LOG_STEP("ut_kvp_openMemory( pInstance, %s - zero length file ) - Negative", KVP_VALID_TEST_ZERO_LENGTH_YAML_FILE);
+    if (read_file(KVP_VALID_TEST_ZERO_LENGTH_YAML_FILE, &pMemoryInstance) == 0)
+    {
+        status = ut_kvp_openMemory(pInstance, pMemoryInstance.buffer, pMemoryInstance.length);
+        UT_ASSERT(status == UT_KVP_STATUS_PARSING_ERROR);
+    }
+
+    /* Positive Tests */
+    UT_LOG_STEP("ut_kvp_openMemory( pInstance,  KVP_VALID_TEST_YAML_FILE ) - Positive");
+    if (read_file(KVP_VALID_TEST_JSON_FILE, &pMemoryInstance) == 0)
+    {
+        status = ut_kvp_openMemory(pInstance, pMemoryInstance.buffer, pMemoryInstance.length);
+        UT_ASSERT(status == UT_KVP_STATUS_SUCCESS);
+    }
+
+    UT_LOG_STEP("ut_kvp_openMemory( pInstance, %s ) - Postive", KVP_VALID_TEST_NOT_VALID_YAML_FORMATTED_FILE);
+    if (read_file(KVP_VALID_TEST_JSON_FILE, &pMemoryInstance) == 0)
+    {
+        status = ut_kvp_openMemory(pInstance, pMemoryInstance.buffer, pMemoryInstance.length);
+        UT_ASSERT(status == UT_KVP_STATUS_SUCCESS);
+    }
+
+    UT_LOG_STEP("ut_kvp_openMemory( pInstance, %s ) - Postive", KVP_VALID_TEST_JSON_FILE);
+    if (read_file(KVP_VALID_TEST_JSON_FILE, &pMemoryInstance) == 0)
+    {
+        status = ut_kvp_openMemory(pInstance, pMemoryInstance.buffer, pMemoryInstance.length);
+        UT_ASSERT(status == UT_KVP_STATUS_SUCCESS);
+    }
+
+    /* Test Destroy causes close this should work fine */
```

Comment:
`destoryInstance` - should have been tested as a seperate test.. You don't need to test it here, you need to seperate your tests more.